Creating the annotated files
============================

Annotated variants
------------------

To run the GWAS analysis the variants in the study need to be annotated with the ld-score and to a gene.
To create the SNPs dataset use the `baghera-tool create-files` command

.. autofunction:: baghera_tool.preprocess.create_files

We use `precomputed ld-score <https://github.com/bulik/ldsc>`_ , from the set of variants for the European population of 1000 Genomes, and  the genes in the `Gencode v31 annotations <https://www.gencodegenes.org/human/>`_ , only the protein coding ones. To cope with overlapping genes, we clustered them, obtaining a dataset of 15000 non-overlapping genes. For the annotation, we use a 50 kb window.
The resulting dataset of annotated variants has around 1.3 millions SNPs, 55% of which are annotated with a gene.

*Please note that this file has already been created, to process the data skip to the next section*

It is possible to annotate a different set of variants, for example another reference panel, using the `create-files` function.
For the moment it only supports .gtf files for the genes annotation and the LD-score folder with the structure in
`<https://github.com/bulik/ldsc>`_

The annotated ld scores table, return as an output, has the following structure:


+-----+------------+-----------+-------+-------+-------+-------+
| chr | rs_id      | position  | cm    | maf   | l     | gene  |
+=====+============+===========+=======+=======+=======+=======+
|9    | rs10123646 | 108998    | 0.090 | 0.499 | 3.155 | FOXD4 |
+-----+------------+-----------+-------+-------+-------+-------+

While the gene table looks like the one below.

+-------+-------+-------+-------+
| chrom | start | stop  | name  |
+=======+=======+=======+=======+
| 1     | 65418 | 71585 | OR4F5 |
+-------+-------+-------+-------+


Create the dataset
------------------

The BAGHERA core analysis uses a table like the one below

+-----+------------+-----------+-------+-------+-------+-------+-------------+-----+
| chr | rs_id      | position  | cm    | maf   | l     | gene  | sample_size | z   |
+=====+============+===========+=======+=======+=======+=======+=============+=====+
|9    | rs10123646 | 108998    | 0.090 | 0.499 | 3.155 | FOXD4 | 361194      | 1.4 |
+-----+------------+-----------+-------+-------+-------+-------+-------------+-----+


Such file is generated by merging  a summary statistics file with the annotated ld file.

To create the SNPs dataset use the `baghera-tool generate-SNPs-file` command

.. autofunction:: baghera_tool.preprocess.generate_snp_file

The function uses a tsv table as input and merges it with the annotated ld score table.

There are different input types managed by the code, specified in the parameter, use the `-t <type>` parameter, with one betweeen
ldsc, ukbb, position, position_ukbb:

- `-t position_ukbb` an **UKBB** file, .assoc.tsv, that can be found in the round two results `here <http://www.nealelab.is/uk-biobank>`_

.. autofunction:: baghera_tool.preprocess.import_position_ukbb

- `-t ldsc` a **sumstats** file, .sumstats.txt that is found on the ldsc file

.. autofunction:: baghera_tool.preprocess.import_ldsc

- `-t position` file

.. autofunction:: baghera_tool.preprocess.import_position

- `-t ukbb` an **UKBB** file, .assoc.tsv. Round one uk biobank results.

.. autofunction:: baghera_tool.preprocess.import_ukbb
